File: src/math/f.ts
import Complex from "complex.js";

const degToRad = (a: number) => a * (Math.PI / 180);

const Ei = Complex(3.17, 1.99 * 10 ** -3);
const Ew = Complex(82, 2.5);

const sin_t = (t: number) => Complex(degToRad(t)).sin();
const cos_t = (t: number) => Complex(degToRad(t)).cos();
const t1 = (t: number) => sin_t(t).div(Ei.sqrt()).asin();

const sin_t1 = (t: number) => t1(t).sin();
const cos_t1 = (t: number) => t1(t).cos();

const f = 1_575_420_000;
const c = 299_792_458;
const k0 = (2 * Math.PI * f) / c;

const rvi_c = (t: number) => Ei.mul(cos_t(t));
const ri_c = (t: number) => Ei.sub(sin_t(t).pow(2)).sqrt();

const rw_c = (t: number) => Ew.sub(Ei.mul(sin_t1(t).pow(2))).sqrt();
const rvw_c1 = (t: number) => Ew.mul(cos_t1(t));
const rvw_c2 = (t: number) => Ei.sqrt().mul(rw_c(t));
const rhw_c = (t: number) => Ei.mul(cos_t1(t));

const rvi = (t: number) =>
  rvi_c(t)
    .sub(ri_c(t))
    .div(rvi_c(t).add(ri_c(t)));

const rhi = (t: number) =>
  cos_t(t)
    .sub(ri_c(t))
    .div(cos_t(t).add(ri_c(t)));

const rvw = (t: number) =>
  rvw_c1(t)
    .sub(rvw_c2(t))
    .div(rvw_c1(t).add(rvw_c2(t)));

const rhw = (t: number) =>
  rhw_c(t)
    .sub(rw_c(t))
    .div(rhw_c(t).add(rw_c(t)));

const e1 = (t: number, d: number) => ri_c(t).mul(d).mul(k0).mul(Complex("2i")).exp();

const Rh = (t: number, d: number) =>
  rhi(t)
    .add(rhw(t).mul(e1(t, d)))
    .div(Complex.ONE.add(rhi(t).mul(rhw(t).mul(e1(t, d)))));

const Rv = (t: number, d: number) =>
  rvi(t)
    .add(rvw(t).mul(e1(t, d)))
    .div(Complex.ONE.add(rvi(t).mul(rvw(t).mul(e1(t, d)))));

const e2 = (t: number, h: number) => cos_t(t).mul(h).mul(k0).mul(Complex("2i")).exp();

const F = (t: number, p0: number, p1: number, p2: number) => p0 + p1 * t + p2 * t ** 2;

export const U = (t: number, h: number, d: number, p0: number, p1: number, p2: number) =>
  F(t, p0, p1, p2) * Complex.ONE.add(Rh(t, d).add(Rv(t, d)).div(2).mul(e2(t, h))).abs();

File: f.py
from cmath import *

ei = 3.2 + 0j
ew = 82 + 2.5j
k0 = (2 * pi * 1_575_420_000) / 299_792_458

def t1(t): return asin(sin(t) / sqrt(ei))
def rvi(t): return (ei * cos(t) - sqrt(ei - sin(t) ** 2)) / (ei * cos(t) + sqrt(ei - sin(t) ** 2))
def rhi(t): return (cos(t) - sqrt(ei - sin(t) ** 2)) / (cos(t) + sqrt(ei - sin(t) ** 2))
def rvw(t): return (ew * cos(t1(t)) - sqrt(ei) * sqrt(ew - ei * sin(t1(t)) ** 2)) / (ew * cos(t1(t)) + sqrt(ei) * sqrt(ew - ei * sin(t1(t)) ** 2))
def rhw(t): return (ei * cos(t1(t)) - sqrt(ew - ei*sin(t1(t)) ** 2)) / (ei * cos(t1(t)) + sqrt(ew - ei*sin(t1(t)) ** 2))
def e(t, d): return exp(2j * k0 *d *sqrt(ei - sin(t) ** 2))
def Rh(t, d): return (rhi(t) + rhw(t) * e(t, d)) / (1 + rhi(t) * rhw(t) * e(t, d))
def Rv(t, d): return (rvi(t) + rvw(t) * e(t, d)) / (1 + rvi(t) * rvw(t) * e(t, d))
def e2(t, h): return exp(2j * k0 * h * cos(t))
def F(t, p0, p1, p2):
    t = t * 180 / pi
    return p0 + p1 * t + p2 * t ** 2
def U(t, d, h, p0, p1, p2): return F(t, p0, p1, p2) * abs(1 + ((Rh(t, d) + Rv(t, d)) / 2) * e2(t, h))


for i in range(91):
    t = i * pi / 180
    print(U(t, 0.5, 1, 1, 1, 1)) 